### ISSUE TITLE
Hard Frame Drop on Mac OS X of Youtube content. 

### ISSUE THREAD

**Issue by appetrosyan** [mpv-player/mpv#8461]
### Important Information

- mpv version 0.33.0
- macOS Version 11.1
- Source of the mpv binary or bundle `brew`
- If known which version of mpv introduced the problem 0.33
- Possible screenshot or video of visual glitches `Dropped: 1283`

> If you're not using git master or the latest release, update.
> Releases are listed here: https://github.com/mpv-player/mpv/releases

I suspect this is related to the old versions of OpenGL and no hardware decoding. What I found is that for some reason an empty `mpv.conf` is not the same as `--no-config`. 

I have tinkered with the smooth video project, but made a clean 
`brew reinstall mpv`

### Reproduction steps

>Try to reproduce your issue with --no-config first. 

Mitigated, but still present. 

> If it isn't reproducible with --no-config try to first find out which option or script causes your issue.

I have an empty `.config/mpv.conf` file, and still see a difference between that and the `--no-config` option.

> Describe the reproduction steps as precise as possible. 

attempt to play a 4k 60fps video from Youtube e.g. 

### Expected behavior
`mpv "https://www.youtube.com/watch?v=nqUPz9nEXPk" --no-config` 

produces the output at a 4k 60fps. 

### Actual behavior
The result is, however, an oversized window, and a large amount of dropped frames. Particularly at the 0:00:37 mark of the video. 

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/5808169/output.txt)

**Comment by Akemi** [comment]
> I have an empty .config/mpv.conf file, and still see a difference between that and the --no-config option.

it is the same or (possibly) you have some other config lingering around somewhere. but can't tell like this.

> If known which version of mpv introduced the problem 0.33

so 0.32 works without problems?

> The result is, however, an oversized window, and a large amount of dropped frames. Particularly at the 0:00:37 mark of the video.

don't mix two/three different problems.
if you don't like the window size use `--autofit=`.
if you have problems with the config, open a separate issue please.

i can't reproduce this and have nearly to no dropped frames. though i am still on 10.14 and probably have a faster system(?). if you have a dedicated GPU try with `--macos-force-dedicated-gpu`. since this is a ~60fps video and it's close to the display refresh rate, it is best to set `--video-sync=display-resample`.

either your mac is too slow for it, apple fucked up opengl even more with macOS 11, we have a regression if earlier versions of mpv work or something else that needs investigating. though in most cases there is most likely not much we can do about.

**Comment by appetrosyan** [comment]
> so 0.32 works without problems?

I think it does. I can't install it on Big Sur, though. 

>either your mac is too slow for it, 

It's a 2016 Pro with AMD RPRO565M. Also `mpv` and `iina`  (mpv + Cocoa GUI) are the only players showing this problem. 

> don't mix two/three different problems. if you don't like the window size use --autofit=. if you have problems with the config, open a separate issue please.

My mistake, I assumed that these may be related.

**Comment by appetrosyan** [comment]
> if you have a dedicated GPU try with `--macos-force-dedicated-gpu`. since this is a ~60fps video and it's close to the display refresh rate, it is best to set `--video-sync=display-resample`.

The two options helped quite a bit. However, this still causes an unusually high load on the system: the fans ramp up and the chassis becomes noticeably hotter. Doesn't happen with Safari. Doesn't happen if I try the same via VLC. 

Here's the updated log. 

[with-dgpu-log.txt](https://github.com/mpv-player/mpv/files/5810093/with-dgpu-log.txt)

**Comment by Akemi** [comment]
> I think it does. I can't install it on Big Sur, though.

what is the problem here? there should be no problem with compatibility. just try with one of the bundles from stolendata https://laboratory.stolendata.net/~djinn/mpv_osx/
this would be an important information to have.

> The two options helped quite a bit. However, this still causes an unusually high load on the system: the fans ramp up and the chassis becomes noticeably hotter. Doesn't happen with Safari. Doesn't happen if I try the same via VLC.

what does 'quite a bit' mean? please try with both options separately. i would expect `--video-sync=display-resample` to be enough at least if it was a timing problem. if it only works well with forcing the dedicated gpu, there is most likely a GPU bottleneck.

that fans ramping up and it getting hotter is expected because it uses the dedicated GPU. 

you can't compare this to safari. first of all safari got vp9 hwdec support (not sure if this is applicable for your system) and it also doesn't utilise opengl. this is by no means comparable.

for VLC it can have different reasons, like the usage of a layer (you could, just for testing, try with `--vo=gpu` and see if it makes better), a rendering pipeline that might be optimised for performance but not correctness/quality of rendering etc. also see #8279

it could also be similar to #5536 and more specific some hardcoded surface size (https://github.com/mpv-player/mpv/issues/5536#issuecomment-370236454) that could decrease performance with the iGPU. does the iGPU actually clock higher when playing back the video or does it stay at low frequencies? how does it compare to VLC, does the clock of the iGPU increase in those cases?

**Comment by appetrosyan** [comment]
> this would be an important information to have.

Tested every release of MPV on stolen data. 

Versions down to 0.24.0 (exclusive) All exhibit same performance with `--no-config`. 

The rest don't run at all. 

> what does 'quite a bit' mean? 

This is a three run average + variance, up to 0:40:00 of linked video. 
`--no-config`                                   1040 ± 20 dropped frames. 
`--video-sync=display-resample` 980 ± 20 dropped frames + Audio Desynchronisation warning
`--macos-force-dedicated-gpu`   20±1 dropped frames + some micro stutter during Logo
both options                                     6 ± 1 dropped frames + no stutter. 
`--vo=gpu` + both options             113 ± 3 dropped frames + no stutter
`--vo=gpu` + `--video-sync`.       959 ± 40 dropped frames + stutter
`--vo=gpu` + `--macos-force...`  72 dropped frames + stutter

>for VLC it can have different reasons, like the usage of a layer (you could, just for testing, try with --vo=gpu and see if it makes better), a rendering pipeline that might be optimised for performance but not correctness/quality of rendering etc. also see #8279. 

VLC ignores 10bit content in other test cases. `--hwdec=auto` makes things worse in terms of microstutter, unless the `--hwdec-extra-frames` is raised (to 255). `--no-hidpi-window-scale` doesn't seem to help things either, though the player starts in a slightly more tenable size. 

The iGPU has a usage spike, but doesn't clock all the way up. I suspect it's thermal throttling, because the fans ramp up in this case too. VLC is consistently at half the GPU (and CPU) utilisation. However, it produces garbled frames at the logo shot up to the next keyframe. It doesn't produce a smooth playback as you would get with the dGPU enabled, though smooth enough to be perceptually above 23.96fps.

**Comment by Akemi** [comment]
then it's at least not a regression.

why do you expect `--no-hidpi-window-scale` to help in any way? it just changes the window size calculation. like i said before could you please leave that unrelated issue out of this one.

> The iGPU has a usage spike, but doesn't clock all the way up. I suspect it's thermal throttling, because the fans ramp up in this case too.

is that the case for `--video-sync=display-resample` and without or does the iGPU clock lower when not using it?

> --hwdec=auto makes things worse in terms of microstutter,...

if it's still about that VP9 sample, then this should not matter at all because ffmpeg does not have vp9 videtoolbox support yet, or are you using a different sample now? anyway we should stay at one sample for now and without hwdec, because that only adds another factor of error and videotoolbox is quite bad anyway.

another thing you could try are the various opengl options, for example `--opengl-pbo` and see if any of those or any combination helps.

i can only assume for now but there are not many possible reasons:
- our rendering piple does too much, and more than the others, which leads to higher utilisation
- we use some functionality, the other don't use, that don't play well with apple's opengl (we would need a workaround)
- we have some dumb logic error somewhere, but that would be one which only affects macOS

in either way we would need to find out which calls cause the high utilisation and/or high wait times, opengl and normal calls. i tried this once but since i can't reproduce this at on my machine it's kinda pointless for me to try and optimise something.

**Comment by appetrosyan** [comment]
> why do you expect --no-hidpi-window-scale to help in any way? it just changes the window size calculation. like i said before could you please leave that unrelated issue out of this one.

I noticed that the framerate seems to scale with the size of the window, even beyond 1:1. I assumed that maybe because of the scaling on my monitor, it was overshooting and rendering at a higher resolution than 2160p and scaling down. If you say it's unrelated I will not mention it. 

>is that the case for --video-sync=display-resample and without or does the iGPU clock lower when not using it?

The margin of error is too large to tell. It appears that this setting makes little to no difference. 

>if it's still about that VP9 sample, then this should not matter at all because ffmpeg does not have vp9 videtoolbox support yet, or are you using a different sample now? anyway we should stay at one sample for now and without hwdec, because that only adds another factor of error and videotoolbox is quite bad anyway.

Yes. I assumed that vp9 had hardware decoding, because I saw a measurable difference (three round average was lower by more than the margin of error). 

> we have some dumb logic error somewhere, but that would be one which only affects macOS

It could be one that affects my particular set-up, and not Mac OS X in general. I'll tinker with some settings, and let you know if it makes it better. 

I have an intuition that this cocoa message
```
2021-01-15 20:45:11.726 mpv[8197:419473] This NSLayoutConstraint is being configured with a constant that exceeds internal limits.  A smaller value will be substituted, but this problem should be fixed. Break on BOOL _NSLayoutConstraintNumberExceedsLimit(void) to debug.  This will be logged only once.  This may break in the future.
```
may be related to the issue, and that the video is rendered at a higher internal resolution and then scaled down to 2160p, but as you said, it is probably unrelated. 

Failing all of that I'll clean wipe the system and see if I can fix it that way. It could be an artefact of home-brew.

**Comment by appetrosyan** [comment]
`oopengl-pbo=yes` seems to help with the dropped frames. The frame rate is still noticeably sub-60, but the iGPU seems to clock higher and for longer.

**Comment by Akemi** [comment]
> I noticed that the framerate seems to scale with the size of the window, even beyond 1:1. I assumed that maybe because of the scaling on my monitor, it was overshooting and rendering at a higher resolution than 2160p and scaling down. If you say it's unrelated I will not mention it.

yes, that can happen. for reproducible test i would suggest fullscreen instead. rendering at a window size the same es the video resolution is most likely the least demanding for scaling, but on the other hand it might allocate bigger textures which in return could lead to a bandwidth problem. 

> I have an intuition that this cocoa message ...

your intuition is most likely wrong. this has nothing to do with rendering and only affects the touch bar support. see issue https://github.com/mpv-player/mpv/issues/7047. though you can test if it changes anything if you disable the touch bar support on build time with `--disable-macos-touchbar`.

> opengl-pbo=yes seems to help with the dropped frames. The frame rate is still noticeably sub-60, but the iGPU seems to clock higher and for longer. With the Dedicated GPU, the playback is smoother, it no longer stutters near the logo, despite mpv reporting more dropped frames in the console the playback seems subjectively smoother.

did you also try this in combination with` --video-sync=display-resample`?


i would like to go some steps back for now, to your first sample `https://www.youtube.com/watch?v=nqUPz9nEXPk`. i would like to have a log for mpv with `--no-config` and one from VLC please, playing back maybe 10-15sec. let me know the dropped frames and the CPU/GPU utilisation on both. a log for VLC can be obtained via `Window > Messages...`. just make sure to open that messages window first before opening the video.

**Comment by Akemi** [comment]
issue #8477 is related to the touch bar, which is quite frank pretty dumb. it's something i wasn't able to fix and i am inclined that it is an Apple bug.

**Comment by Akemi** [comment]
this most likely was either the same problem that was fixed with these commits or a general opengl problem:
https://github.com/mpv-player/mpv/commit/7f8d069087ab31cfe502e6daad4e04a37399a125
https://github.com/mpv-player/mpv/commit/aa00ad06aae264796a84327f1716534b4f37c622
PR: https://github.com/mpv-player/mpv/pull/8486

we have the macvk gpu-next backend now and performance problems should be re-evaluated now.
